src/mapping_igvc/src/ObstacleManager.cpp

```diff
@@ -0,0 +1,328 @@
+    // is inclusive, and so will return the endpoint when given the range: [1,1]
+    if (max_u != 1){
+        points_after_new_line = current_line.getInterpolationPointsInRange(max_u, 1);
+    }
+
+    // Get the interpolation points from the new line
+    std::vector<Point2D> points_on_new_line = new_line.getInterpolationPointsInRange(0,1);
+
+    // Check if the points of the new line are in the "opposite" order of those
+    // in the known line. That is, is the closest point to the start of the new
+    // line after the closest point to the end of the new line on the old line?
+    if (closest_point_to_start > closest_point_to_end){
+        std::reverse(points_on_new_line.begin(), points_on_new_line.end());
+    }
+
+    // Merge the 3 sets of interpolation points in order:
+    // - points on the current line before the new line
+    // - points on the new line
+    // - points on the current line after the new line
+    std::vector<Point2D> new_interpolation_points = points_before_new_line;
+    new_interpolation_points.insert(
+            new_interpolation_points.end(), points_on_new_line.begin(), points_on_new_line.end());
+    new_interpolation_points.insert(
+            new_interpolation_points.end(), points_after_new_line.begin(), points_after_new_line.end());
+
+    // Return a spline interpolated through our new points
+    return Spline(new_interpolation_points);
+}
+
+nav_msgs::OccupancyGrid ObstacleManager::generateOccupancyGrid() {
+
+    // TODO: maybe prune obstacles outside a given distance from us? (might want a seperate function for this)
+
+    // Find what cells are directly occupied (ie. overlapping) with
+    // known obstacles
+    std::vector<Point2D> occupied_points;
+
+    for (auto& line : lines){
+        // Figure out how many points to sample
+        auto num_sample_points = (int)std::ceil(line.approxLength() / occ_grid_cell_size) * 2;
+
+        // Sample points from the line
+        for (int i = 0; i <= num_sample_points; i++){
+            occupied_points.emplace_back(line(i / (double)num_sample_points));
+        }
+    }
+
+    for (auto& cone : cones){
+        // Iterate over the cone in 2D
+        auto cone_radius_num_cells = (int)std::ceil(cone.radius / occ_grid_cell_size);
+        for (int x = -cone_radius_num_cells; x <= cone_radius_num_cells; x++){
+            for (int y = -cone_radius_num_cells; y <= cone_radius_num_cells; y++) {
+                // Check that the point is within the circle by using the equation of
+                // a circle: (x^2 + y^2 = r^2)
+                if ((std::pow(x, 2) + std::pow(y,2)) <= std::pow(cone_radius_num_cells,2)){
+                    occupied_points.emplace_back(Point2D(
+                            cone.center.x + x * occ_grid_cell_size,
+                            cone.center.y + y * occ_grid_cell_size
+                    ));
+                }
+            }
+        }
+    }
+
+    nav_msgs::OccupancyGrid occ_grid;
+
+    // TODO: Set header
```

Comment:
Speaking of headers, does this account the different frames the `cone` and `line` obstacles live in? I'm assuming it will publish it in the `odom_combined` frame (i.e. relative to the robot).

Reply:
I'm hypothesizing that we'll have to convert all the cones and lines to the `map` frame and store it in this frame as well.

Reply:
After a bit more work, I have things setup such that we can set the output frame of the occupancy grid as a rosparam. We will have to translate incoming obstacles, but I'm planning to do this in the node itself so that we can test the `ObstacleManager` class without having to run ROS.